keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+### Non-Goals
+
+1. Support anything other than KRM-style plugins that follow the [functions spec](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md)
+1. Directly implement capabilities to publish plugin providers or other resources to OCI registries 
+
+## Proposal
+
+In order to standardize plugin discovery, introduce a `Catalog` API `kind` recognized by `kustomize build`. One or more `Catalog` resources can be referenced from any Kustomize kind as either a local file or remote reference available as from an HTTP(s) endpoint or as an [OCI artifact](https://github.com/opencontainers/artifacts).
+
+A `Catalog` will contain a collection of one or more modules that can be used with a Kustomize resource. 
+
+A minimal example is shown below:
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+```
+
+This will enable a users to use plugin providers with provider information omitted, such as:
+
+```yaml
+# app/kustomization.yaml
+kind: Kustomization
+catalogs:
+  - https://example.com/plugins/catalog.yaml
+resources:
+- javaapp.yaml
+```
+
+```yaml
+# app/javaapp.yaml
+apiVersion: example.com/v1
+kind: JavaApplication
+spec:
+  application: team/my-app
+  version: v1.0
+```
+
+When this Kustomization is processed by `kustomize build`, the referenced catalog (or catalogs) will be used to locate a plugin provider that supports the apiVersion `example.com/v1` and kind `JavaApplication`. If found in one of the referenced catalogs, kustomize can determine the provider configuration without the need for the user to specify it in the kustomization resources directly. The catalogs will be searched in order specified. If more than one catalog defines the target apiVersion and kind, the first will be selected. 
+
+In addition to the new catalog kind, `kustomize build` will accept a repeatable flag `--trusted-plugin-catalog=""`. When present, this flag instructs `kustomize build` to automatically fetch and execute plugins that are defined by the catalog and referenced within the Kustomization, Component or Composition. When a resource is processed by `kustomize build` and a catalog is referenced but not specified using the `--trusted-plugin-catalog=""` flag, an error will occur. Kustomize will provide a built in Catalog for supporting official extensions, published to a well publicized endpoint. This catalog will _NOT_ require the user to explicitly trust it. Users can provide the `apiVersion` and `kind` of the official extensions in kustomize resource and these will be resolved by the official catalog.
```

Comment:
Suggestion: We should consider adding a command like `kustomize view catalog` that shows the catalogs trusted vs. required for the current configuration object, along with publisher information retrieved from the catalog itself.

> Minor nit, I find --trusted-plugin-catalog="" a bit verbose. Is just --catalog="" not sufficient?

That verbosity is coming from my input. The idea is to make sure the user is aware that using this flag is a declaration of trust in the plugins the referenced file includes. I feel like `--catalog=` could be interpreted as "I need this" rather than "I trust this". I could be convinced that "plugin" is unnecessary though. The whole purpose of the flag as opposed to the inline field is for the user to prove to us that they know what they're doing, the same way they do by installing the plugin to a particular location themselves in the legacy flow.

> Will this builtin catalog have docker images/starlark functions?

Totally up to us, and I think that decision doesn't need to be in scope of the KEP. FWIW I think we would want to use docker containers (e.g. your new helm transformer implementation), but not starlark.